core/src/main/java/hivemall/smile/classification/MaxEntUDTF.java

```diff
@@ -0,0 +1,440 @@
+        int label = PrimitiveObjectInspectorUtils.getInt(args[1], labelOI);
+        labels.add(label);
+    }
+    
+    private void parseFeatures(@Nonnull final Object argObj, @Nonnull final MatrixBuilder builder) {
+    	final int length = featureListOI.getListLength(argObj);
+        for (int i = 0; i < length; i++) {
+            Object o = featureListOI.getListElement(argObj, i);
+            if (o == null) {
+                continue;
+            }
+            double v = PrimitiveObjectInspectorUtils.getDouble(o, featureElemOI);
+            builder.nextColumn(i, v);
+        } 
+        builder.nextRow();
+    }
+    
+    @Override
+    public void close() throws HiveException {
+        this._progressReporter = getReporter();
+        this._treeBuildTaskCounter = (_progressReporter == null) ? null
+                : _progressReporter.getCounter("hivemall.smile.MaxEntClassifier$Counter",
+                    "finishedGISTask");
+        reportProgress(_progressReporter);
+
+        if (!labels.isEmpty()) {
+            Matrix x = matrixBuilder.buildMatrix();
+            this.matrixBuilder = null;
+            int[] y = labels.toArray();
+            this.labels = null;
+
+            // run training
+            train(x, y);
+        }
+
+        // clean up
+        this.featureListOI = null;
+        this.featureElemOI = null;
+        this.labelOI = null;
+    }
+    
+    private void checkOptions() throws HiveException {
+    	if (_USE_SMOOTHING == false && _SMOOTHING_OBSERVATION != 0.1) {
+            throw new HiveException("Instructions received to avoid smoothing, but smoothing constant is set [" + _SMOOTHING_OBSERVATION + "]");
+        }
+    }
+    
+    /**
+     * @param x features
+     * @param y label
+     * @param attrs attribute types
+     * @param numTrees The number of trees
+     * @param numVars The number of variables to pick up in each node.
+     * @param seed The seed number for Random Forest
+     */
+    private void train(@Nonnull Matrix x, @Nonnull final int[] y) throws HiveException {
+        final int numExamples = x.numRows();
+        if (numExamples != y.length) {
+            throw new HiveException(String.format("The sizes of X and Y don't match: %d != %d",
+                numExamples, y.length));
+        }
+        checkOptions();
+
+        int[] labels = SmileExtUtils.classLables(y);
+        Attribute[] attributes = SmileExtUtils.attributeTypes(_attributes, x);
+
+        if (logger.isInfoEnabled()) {
+            logger.info("real: " + _real + ", smoothing: " + this._USE_SMOOTHING + ", smoothing constant: "
+                    + _SMOOTHING_OBSERVATION);
+        }
+
+        IntMatrix prediction = new DoKIntMatrix(numExamples, labels.length); // placeholder for out-of-bag prediction
+        AtomicInteger remainingTasks = new AtomicInteger(_numTrees);
+        List<TrainingTask> tasks = new ArrayList<TrainingTask>();
+        for (int i = 0; i < _numTrees; i++) {
```

Comment:
I left an ability to divide training set further into subsets for memory reasons. Perhaps at times it is faster to subdivide the data that mapper accepts into a number of chunks and train a model per each training set chunk. Implementation-wise it is same to Random Forests indeed. At the moment _numTrees is set to 1. If it works with 1 all the time then perhaps it is a good setting, otherwise it may be better to allow people to set it.